parser: scan and split statements before parsing, return SQL strings for statements

RaduBerinde opened the pull request:
These series of changes move the handling of `;` out of the parser and into a pre-processing stage (which scans lexical tokens and splits statements). The Parser now returns the strings for the statements, which are then used instead of `String()`. Eventually we will cache ASTs and these strings will be used to look up ASTs in the cache and avoid parsing entirely.

Functionally, only two small things change (both positive IMO):
 - traces and logs now show the original SQL instead of the roundtripped string
 - when we hit a parsing error in a multi-statement, we now show only the relevant statement (this is what postgres shows too).

In principle, this should be less efficient because we need to remember the scanned tokens. However, I made some improvements that offset the overhead. Also note that the code that handled multiple statements in sql.y was inefficient (every prefix of the `[]tree.Statement` slice was being copied on heap when we assigned it to the `val interface{}`).

#### parser: split lexer

This change splits the lexer into two parts:
 - `scanner` does the low-level scanning of the tokens
 - `lexer` operates on a slice of tokens, performs lookahead and
   handles errors

The separation of these steps will allow us to split statements and
parse each one individually.

Release note: None

#### parser: split statements prior to parsing

This change uses the scanner to split the input into statements and
then parses each one independently. This will allow the caller to get
the exact SQL for each statement. The SQL grammar does not allow for
semicolons anymore.

Release note: None

#### parser: make sqlSymType smaller

Change `id` and `pos` to `int32` to make the structure smaller.

Release note: None

#### parser: return SQL strings alongside the statements

Release note: None




Benchmarks:
```
name                 old time/op    new time/op    delta
Parse/simple           11.8µs ± 5%     9.5µs ± 1%  -19.22%  (p=0.000 n=12+12)
Parse/string           15.8µs ± 3%    13.6µs ± 1%  -13.95%  (p=0.000 n=12+12)
Parse/tpcc-delivery    19.8µs ± 3%    19.0µs ± 2%   -4.18%  (p=0.000 n=12+12)
Parse/account          34.0µs ± 2%    33.8µs ± 4%     ~     (p=0.101 n=12+12)

name                 old alloc/op   new alloc/op   delta
Parse/simple           2.44kB ± 0%    2.52kB ± 0%   +3.28%  (p=0.000 n=12+12)
Parse/string           2.71kB ± 0%    3.56kB ± 0%  +31.33%  (p=0.000 n=12+12)
Parse/tpcc-delivery    3.10kB ± 0%    5.49kB ± 0%  +76.80%  (p=0.000 n=12+12)
Parse/account          5.81kB ± 0%    9.06kB ± 0%  +55.92%  (p=0.000 n=12+12)

name                 old allocs/op  new allocs/op  delta
Parse/simple             23.0 ± 0%      21.0 ± 0%   -8.70%  (p=0.000 n=12+12)
Parse/string             27.0 ± 0%      26.0 ± 0%   -3.70%  (p=0.000 n=12+12)
Parse/tpcc-delivery      35.0 ± 0%      35.0 ± 0%     ~     (all equal)
Parse/account            81.0 ± 0%      81.0 ± 0%     ~     (all equal)
```

RaduBerinde committed:
parser: split statements prior to parsing

This change uses the scanner to split the input into statements and
then parses each one independently. This will allow the caller to get
the exact SQL for each statement. The SQL grammar does not allow for
semicolons anymore.

Release note: None

RaduBerinde committed:
parser: split lexer

This change splits the lexer into two parts:
 - `scanner` does the low-level scanning of the tokens
 - `lexer` operates on a slice of tokens, performs lookahead and
   handles errors

The separation of these steps will allow us to split statements and
parse each one individually.

Release note: None

RaduBerinde committed:
parser: return SQL strings alongside the statements

Also, we revert `ExecStmt/PrepareStmt/Statement` to using
`Stmt.String()` instead of `SQL` for stringification, since the former
obfuscates passwords. The `SQL` field will be used for the query plan
cache.

Release note: None

RaduBerinde committed:
parser: make sqlSymType smaller

Change `id` and `pos` to `int32` to make the structure smaller.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/32982)
<!-- Reviewable:end -->

RaduBerinde commented:
![image](https://user-images.githubusercontent.com/16544120/49747740-18f54a80-fc72-11e8-86b4-de51138e6c37.png)

The end is near!

mjibson commented on pkg/sql/pgwire/conn.go:
```diff
@@ -490,21 +490,10 @@ func (c *conn) handleSimpleQuery(
-		} else {
-			// TODO(radu): return this information from the parser instead of using
-			// String.
-			sqlStr = stmt.String()
```
This used the FmtSimple context which hides passwords and maybe other stuff by default. It looks like this change remove that which means things that were previously scrubbed on log files are no longer scrubbed.

madelynnblue reviewed (approved):
LGTM but @knz should be given time to chime in about the string obfuscation stuff in case I'm wrong or it's missing somewhere else.

knz commented:
this may have to wait until tomorrow night as I have a couple of hot potatoes to address before then

RaduBerinde commented:
Well, I hit a lot of merge conflicts with #32848 :/  I might have to squash the commits to deal with it.

RaduBerinde commented:
Never mind, I was able to keep the commits.

@bobvawter it would be great if you could take a look to double-check I didn't screw up your change.

bobvawter commented:
@RaduBerinde, thanks for the heads-up.  Your changes look entirely compatible with #32848.  There are only the two places in `conn.go` where we allow the int size to be changed.

RaduBerinde commented:
@knz let me know if you still plan to take a look or I can merge.

knz reviewed (commented):
The code is functionally correct but:

1) I'd like to see a before/after benchmark with `-benchmem` to see how the alloc profile is changed.
2) you need a conditional somewhere to free (enable GC of) the token buffer after you're done parsing a very long statement (we have some client statements that pass VALUE clauses with thousands of multi-col entries), so that the tok buf doesn't stay around forever in the session. Maybe release the tokenbuf if it has grown larger than 1000 tokens in the previous parse?
3) please check, because I don't remember: if you resize a Go slice from N to 0 elements without changing its capacity, does that enable the GC to free all remaining elements in the range [1:N-1]? If not, your PR as-is is creating a potentially large memory leak.

> Reviewed 7 of 7 files at r1, 8 of 8 files at r2, 5 of 5 files at r3, 21 of 21 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32982#-:-LTdCspH1JXytmtGYegP:bzayd5f)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/parser/parse.go, line 270 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32982#-LTdCiYM9_Xnz83Mmrcm:-LTdCiYM9_Xnz83Mmrcn:bmzjhb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b04d8423cd1fd6482500aa69f035a463751c5475/pkg/sql/parser/parse.go#L270)):*
> ```Go
> }
> 
> //func (p *Parser) SplitStmts(sql string) []string {
> ```

I don't see any commit where SplitStmts is used -- if it's dead code you may want to remove it from the intermediate commits too.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> I'd like to see a before/after benchmark with -benchmem to see how the alloc profile is changed.

Then scroll up :)

2 and 3 are very good points, thank you very much! I'll find a way to fix that. On 3, I don't think they get freed, you can actually get them back by reslicing (eg https://play.golang.org/p/hBHkYahoo0y).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32982#-:-LTdHMEo3SV7steX6nQo:b-dzukfx)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/parser/parse.go, line 270 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/32982#-LTdCiYM9_Xnz83Mmrcm:-LTdHH0uAp5efsDSlkAK:bwcoxo8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b04d8423cd1fd6482500aa69f035a463751c5475/pkg/sql/parser/parse.go#L270)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't see any commit where SplitStmts is used -- if it's dead code you may want to remove it from the intermediate commits too.
</blockquote></details>

Ah, I removed it in the wrong commit.


<!-- Sent from Reviewable.io -->

knz commented:
yeah I found the benchmark thanks

LGTM when you have addressed the other two things

RaduBerinde commented:
Ha ha, it looks like we don't actually reuse a Parser object. We should fix that though.

RaduBerinde committed:
sql: reuse parser

Instead of allocating a `Parser` object for each statement, embed one
in `pgwire.conn` and reuse it. Also, remove the unused parser embedded
in the planner.

The parser code is fixed up to better handle reuse: `scanner` and
`lexer` now have cleanup methods to release any memory they might be
holding on to.

Release note: None

RaduBerinde commented:
Rebased and added a commit that cleans up the scanner and lexer after they are used, and adds a Parser in `pgwire.conn` that can be reused.

mjibson commented on pkg/sql/parser/lexer.go:
```diff
@@ -54,6 +54,14 @@ func (l *lexer) init(
+// cleanup is used to avoid holding on to memory unnecessarily (for the cases
+// where we reuse a scanner).
+func (l *lexer) cleanup() {
+	l.tokens = nil
```
Does it make sense to do something like `if cap(l.tokens) > 1000 { l.tokens = nil } else { l.tokens = l.tokens[:0] }` to avoid reallocating on repeated statements?

RaduBerinde replied:
The lexer doesn't allocate data, this is just an input to it. The slice comes from the scanner, and we initialize that with a preallocated buffer embedded in the parser. We could reuse that if it's reallocated, but if we're ok with holding on to a bigger buffer forever we should just increase the size of the embedded buffer.

knz reviewed (approved):
> Reviewed 3 of 8 files at r6, 3 of 5 files at r7, 17 of 21 files at r8, 7 of 7 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/32982)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
